lib/posix-socket/socket_vnops.c

```diff
@@ -0,0 +1,323 @@
+
+static uint64_t s_inode = 0;
+
+#define posix_socket_getattr ((vnop_getattr_t) vfscore_vop_einval)
+#define posix_socket_inactive ((vnop_inactive_t) vfscore_vop_nullop)
+
+/* vnode operations */
+struct vnops posix_socket_vnops = {
+	.vop_close = posix_socket_vfscore_close,
+	.vop_write = posix_socket_vfscore_write,
+	.vop_read = posix_socket_vfscore_read,
+	.vop_ioctl = posix_socket_vfscore_ioctl,
+	.vop_getattr = posix_socket_getattr,
+	.vop_inactive = posix_socket_inactive,
+};
+
+#define posix_socket_vget ((vfsop_vget_t) vfscore_nullop)
+
+/* file system operations */
+static struct vfsops posix_socket_vfsops = {
+	.vfs_vget = posix_socket_vget,
+	.vfs_vnops = &posix_socket_vnops,
+};
+
+/* bogus mount point used by all sockets */
+static struct mount posix_socket_mount = {
+	.m_op = &posix_socket_vfsops
+};
+
+struct posix_socket_file *
+posix_socket_file_get(int sock_fd)
+{
+	struct posix_socket_file *file = NULL;
+	struct vfscore_file *fos;
+
+	fos = vfscore_get_file(sock_fd);
+
+	if (!fos) {
+		uk_pr_err("Failed with invalid descriptor\n");
+		file = ERR2PTR(EINVAL);
+		goto EXIT;
+	}
+
+	if (fos->f_dentry->d_vnode->v_type != VSOCK) {
+		uk_pr_err("File descriptor is not a socket\n");
+		file = ERR2PTR(EINVAL);
+		goto EXIT;
+	}
+
+	file = fos->f_data;
+
+EXIT:
+	return file;
+}
+
+int
+socket_alloc_fd(struct posix_socket_driver *d, int type, void *sock_data)
+{
+	int ret = 0;
+	int vfs_fd;
+	struct posix_socket_file *file = NULL;
+	struct vfscore_file *vfs_file = NULL;
+	struct dentry *s_dentry;
+	struct vnode *s_vnode;
+
+	/* Reserve file descriptor number */
+	vfs_fd = vfscore_alloc_fd();
+	if (vfs_fd < 0) {
+		ret = -ENFILE;
+		uk_pr_err("failed to allocate file descriptor number\n");
+		goto ERR_EXIT;
+	}
+
+	/* Allocate file, dentry, and vnode */
+	file = uk_calloc(d->allocator, 1, sizeof(*file));
+	if (!file) {
+		ret = -ENOMEM;
+		uk_pr_err("failed to allocate socket file\n");
+		goto ERR_MALLOC_FILE;
+	}
+
+	vfs_file = uk_calloc(d->allocator, 1, sizeof(*vfs_file));
+	if (!vfs_file) {
+		ret = -ENOMEM;
+		uk_pr_err("failed to allocate socket vfs_file\n");
+		goto ERR_MALLOC_VFS_FILE;
+	}
+
+	ret = vfscore_vget(&posix_socket_mount, s_inode++, &s_vnode);
+	UK_ASSERT(ret == 0); /* we should not find it in cache */
+
+	if (!s_vnode) {
+		ret = -ENOMEM;
+		uk_pr_err("failed to allocate socket vnode\n");
+		goto ERR_ALLOC_VNODE;
+	}
+
+	uk_mutex_unlock(&s_vnode->v_lock);
+
+	/*
+	 * it doesn't matter that all the dentries have the
+	 * same path since we never lookup for them
+	 */
+	s_dentry = dentry_alloc(NULL, s_vnode, "/");
+
+	if (!s_dentry) {
+		ret = -ENOMEM;
+		uk_pr_err("failed to allocate socket dentry\n");
+		goto ERR_ALLOC_DENTRY;
+	}
+
+	/* Put things together, and fill out necessary fields */
+	vfs_file->fd = vfs_fd;
+	vfs_file->f_flags = UK_FWRITE | UK_FREAD;
+	vfs_file->f_count = 1;
+	vfs_file->f_data = file;
+	vfs_file->f_dentry = s_dentry;
+	vfs_file->f_vfs_flags = UK_VFSCORE_NOPOS;
+
+	// s_vnode->v_data = file;
```

Comment:
Just a nitpick, but I would drop this line.

Reply:
This line should not be commented out, turns out it is very important and an artifact of previous debugging :)